kite-tools-parent/kite-tools/src/main/java/org/kitesdk/cli/commands/TarImportCommand.java

```diff
@@ -0,0 +1,167 @@
+    Preconditions.checkArgument(targets != null && targets.size() == 2,
+        "Tar path and target dataset name are required.");
+
+    Path source = qualifiedPath(targets.get(0));
+    FileSystem sourceFS = source.getFileSystem(getConf());
+    Preconditions.checkArgument(sourceFS.exists(source),
+        "Tar path does not exist: " + source);
+
+    Preconditions.checkArgument(SUPPORTED_TAR_COMPRESSION_TYPES.contains(compressionType),
+        "Compression type " + compressionType + " is not supported");
+
+    String datasetName = targets.get(1);
+
+    int success = 0;
+
+    DatasetRepository repo = getDatasetRepository();
+
+    Dataset<GenericRecord> tarImport = null;
+    if (!repo.exists(namespace, datasetName)) {
+      console.info("Creating dataset {} in repo {}", datasetName, repo.getUri());
+      tarImport = repo.create(namespace, datasetName,
+          new DatasetDescriptor.Builder()
+              .schema(new Schema.Parser().parse(
+                  "{\n" +
+                      "  \"type\": \"record\",\n" +
+                      "  \"name\": \"tarimport\",\n" +
+                      "  \"fields\": [\n" +
+                      "  {\n" +
+                      "    \"type\": \"string\",\n" +
+                      "    \"name\": \"filename\"\n" +
+                      "  },\n" +
+                      "  {\n" +
+                      "    \"type\": \"bytes\",\n" +
+                      "    \"name\": \"filecontent\"\n" +
+                      "  }\n" +
+                      "  ]\n" +
+                      "}"
+              ))
+              .build()
+      );
+    } else {
+      console.info("Using existing dataset {} at {}", datasetName, repo.getUri());
+      tarImport = repo.load(namespace, datasetName);
+    }
+
+    DatasetWriter<GenericRecord> writer = tarImport.newWriter();
+
+    // Create a Tar input stream wrapped in appropriate decompressor
+    // Enhancement would be to use native compression libs
+    TarInputStream tis = null;
+    if (compressionType.equals("gzip")) {
+      tis = new TarInputStream(new GzipCompressorInputStream(sourceFS.open(source)));
+    } else if (compressionType.equals("bzip2")) {
+      tis = new TarInputStream(new BZip2CompressorInputStream(sourceFS.open(source)));
+    } else {
+      tis = new TarInputStream(sourceFS.open(source));
+    }
+    TarEntry entry = null;
+    GenericRecordBuilder builder = new GenericRecordBuilder(tarImport.getDescriptor().getSchema());
+
+    try {
+      int count = 0;
+      boolean readError = false;
+      while ((entry = tis.getNextEntry()) != null && !readError) {
+        if (!entry.isDirectory()) {
+          long size = entry.getSize();
+          byte[] buf = new byte[(int) size];
+          int read = 0;
+          int br = 0;
+          // TarInputStream does not always read what you ask for so loop until it does
+          // Note that this requires the entire tar entry (not entire file) to fit into memory
```

Comment:
I think it probably makes some sense to limit the size. Avro blocks (if I remember correctly) are about 1MB, but records are never split across blocks. That means this will probably stress Avro's assumptions about the data it stores. That should be okay within reason, but I'd like to avoid the situation where the Avro block starts getting within an order of magnitude of the HDFS block size. That would degrade the performance when reading from input splits.

I realize it's a lot more work and would require some custom FS code to build a large file store inside the dataset, so I don't think this is a blocker to get the feature in.

It would be great to get @tomwhite's thoughts on this as well.